File: backend/app/services/brightdata_client.py
"""Thin wrapper around Bright Data's REST APIs.

Function names mirror the Bright Data MCP tool names 1:1 (see bright-data.md)
so this module can be swapped for a real MCP client later without touching
callers. Two request shapes:

- Web Unlocker / SERP: synchronous, POST /request
- Datasets (web_data_* structured extractors): async trigger -> poll -> download
"""

import time

import httpx

from app.config import settings

BASE_URL = "https://api.brightdata.com"

# Per-platform dataset IDs (format gd_xxxxxxxxxxxx), looked up from the
# Bright Data dashboard (Scraper Configuration tab). Fill in once the
# account is provisioned -- see "Open items" in bright-data.md.
DATASET_IDS: dict[str, str] = {
    "web_data_linkedin_person_profile": "",
    "web_data_linkedin_people_search": "",
    "web_data_linkedin_company_profile": "",
    "web_data_linkedin_posts": "",
    "web_data_crunchbase_company": "",
    "web_data_reddit_posts": "",
}


def _headers() -> dict:
    return {
        "Authorization": f"Bearer {settings.brightdata_api_token}",
        "Content-Type": "application/json",
    }


def fetch_url(url: str, *, zone: str | None = None, format: str = "raw") -> str:
    """Web Unlocker: fetch a single URL, bot-protection handled server-side."""
    zone = zone or settings.brightdata_web_unlocker_zone
    with httpx.Client(timeout=60) as client:
        resp = client.post(
            f"{BASE_URL}/request",
            headers=_headers(),
            json={"zone": zone, "url": url, "format": format},
        )
        resp.raise_for_status()
        return resp.text


def scrape_as_markdown(url: str) -> str:
    return fetch_url(url, format="raw")


def search_engine(query: str, *, engine: str = "google", zone: str | None = None) -> str:
    """SERP API: search results as parsed JSON (brd_json=1)."""
    from urllib.parse import quote

    engine_urls = {
        "google": f"https://www.google.com/search?q={quote(query)}&brd_json=1",
        "bing": f"https://www.bing.com/search?q={quote(query)}&brd_json=1",
    }
    return fetch_url(engine_urls[engine], zone=zone, format="raw")


def trigger_dataset(dataset_id: str, inputs: list[dict]) -> str:
    """Kick off an async Datasets collection job. Returns a snapshot_id."""
    with httpx.Client(timeout=30) as client:
        resp = client.post(
            f"{BASE_URL}/datasets/v3/trigger",
            headers=_headers(),
            params={"dataset_id": dataset_id},
            json=inputs,
        )
        resp.raise_for_status()
        return resp.json()["snapshot_id"]


def poll_snapshot(snapshot_id: str, *, timeout: float = 120, interval: float = 3) -> str:
    """Poll until the snapshot is ready/failed, or raise TimeoutError."""
    with httpx.Client(timeout=30) as client:
        elapsed = 0.0
        while elapsed < timeout:
            resp = client.get(f"{BASE_URL}/datasets/v3/progress/{snapshot_id}", headers=_headers())
            resp.raise_for_status()
            status = resp.json().get("status")
            if status in ("ready", "failed"):
                return status
            time.sleep(interval)
            elapsed += interval
    raise TimeoutError(f"snapshot {snapshot_id} not ready after {timeout}s")


def get_snapshot_data(snapshot_id: str) -> list[dict]:
    with httpx.Client(timeout=60) as client:
        resp = client.get(f"{BASE_URL}/datasets/v3/snapshot/{snapshot_id}", headers=_headers())
        resp.raise_for_status()
        return resp.json()


def run_dataset_tool(tool_name: str, inputs: list[dict], *, timeout: float = 120) -> list[dict]:
    """Convenience: trigger + poll + download for a web_data_* tool."""
    dataset_id = DATASET_IDS.get(tool_name)
    if not dataset_id:
        raise ValueError(f"no dataset_id configured for {tool_name!r} -- set it in DATASET_IDS")
    snapshot_id = trigger_dataset(dataset_id, inputs)
    status = poll_snapshot(snapshot_id, timeout=timeout)
    if status != "ready":
        raise RuntimeError(f"{tool_name} snapshot {snapshot_id} ended with status={status}")
    return get_snapshot_data(snapshot_id)

File: backend/src/services/brightdataClient.ts
/**
 * Thin wrapper around Bright Data's REST APIs.
 *
 * Function names mirror the Bright Data MCP tool names 1:1 (see bright-data.md)
 * so this module can be swapped for a real MCP client later without touching
 * callers. Two request shapes:
 *
 * - Web Unlocker / SERP: synchronous, POST /request
 * - Datasets (web_data_* structured extractors): async trigger -> poll -> download
 */

import { config } from "../config.js";

const BASE_URL = "https://api.brightdata.com";

/** Per-platform dataset IDs (format gd_xxxxxxxxxxxx), looked up from the
 * Bright Data dashboard (Scraper Configuration tab). Fill in once the
 * account is provisioned -- see "Open items" in bright-data.md. */
export const DATASET_IDS: Record<string, string> = {
  web_data_linkedin_person_profile: "",
  web_data_linkedin_people_search: "",
  web_data_linkedin_company_profile: "",
  web_data_linkedin_posts: "",
  web_data_crunchbase_company: "",
  web_data_reddit_posts: "",
};

function headers(): Record<string, string> {
  return {
    Authorization: `Bearer ${config.brightdataApiToken}`,
    "Content-Type": "application/json",
  };
}

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

async function ensureOk(response: Response, what: string): Promise<Response> {
  if (!response.ok) {
    throw new Error(`${what} failed: ${response.status} ${response.statusText}`);
  }
  return response;
}

/** Web Unlocker: fetch a single URL, bot-protection handled server-side. */
export async function fetchUrl(
  url: string,
  { zone, format = "raw" }: { zone?: string; format?: string } = {},
): Promise<string> {
  const response = await fetch(`${BASE_URL}/request`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({ zone: zone ?? config.brightdataWebUnlockerZone, url, format }),
    signal: AbortSignal.timeout(60_000),
  });
  await ensureOk(response, "web unlocker request");
  return response.text();
}

export function scrapeAsMarkdown(url: string): Promise<string> {
  return fetchUrl(url, { format: "raw" });
}

/** SERP API: search results as parsed JSON (brd_json=1). */
export function searchEngine(
  query: string,
  { engine = "google", zone }: { engine?: "google" | "bing"; zone?: string } = {},
): Promise<string> {
  const engineUrls = {
    google: `https://www.google.com/search?q=${encodeURIComponent(query)}&brd_json=1`,
    bing: `https://www.bing.com/search?q=${encodeURIComponent(query)}&brd_json=1`,
  };
  return fetchUrl(engineUrls[engine], { zone, format: "raw" });
}

/** Kick off an async Datasets collection job. Returns a snapshot_id. */
export async function triggerDataset(datasetId: string, inputs: unknown[]): Promise<string> {
  const response = await fetch(`${BASE_URL}/datasets/v3/trigger?dataset_id=${datasetId}`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify(inputs),
    signal: AbortSignal.timeout(30_000),
  });
  await ensureOk(response, "dataset trigger");
  const body = (await response.json()) as { snapshot_id: string };
  return body.snapshot_id;
}

/** Poll until the snapshot is ready/failed, or throw. */
export async function pollSnapshot(
  snapshotId: string,
  { timeoutMs = 120_000, intervalMs = 3_000 }: { timeoutMs?: number; intervalMs?: number } = {},
): Promise<string> {
  let elapsed = 0;
  while (elapsed < timeoutMs) {
    const response = await fetch(`${BASE_URL}/datasets/v3/progress/${snapshotId}`, {
      headers: headers(),
      signal: AbortSignal.timeout(30_000),
    });
    await ensureOk(response, "dataset progress");
    const { status } = (await response.json()) as { status?: string };
    if (status === "ready" || status === "failed") return status;
    await sleep(intervalMs);
    elapsed += intervalMs;
  }
  throw new Error(`snapshot ${snapshotId} not ready after ${timeoutMs}ms`);
}

export async function getSnapshotData(snapshotId: string): Promise<unknown[]> {
  const response = await fetch(`${BASE_URL}/datasets/v3/snapshot/${snapshotId}`, {
    headers: headers(),
    signal: AbortSignal.timeout(60_000),
  });
  await ensureOk(response, "dataset snapshot download");
  return (await response.json()) as unknown[];
}

/** Convenience: trigger + poll + download for a web_data_* tool. */
export async function runDatasetTool(
  toolName: string,
  inputs: unknown[],
  { timeoutMs = 120_000 }: { timeoutMs?: number } = {},
): Promise<unknown[]> {
  const datasetId = DATASET_IDS[toolName];
  if (!datasetId) {
    throw new Error(`no dataset_id configured for '${toolName}' -- set it in DATASET_IDS`);
  }
  const snapshotId = await triggerDataset(datasetId, inputs);
  const status = await pollSnapshot(snapshotId, { timeoutMs });
  if (status !== "ready") {
    throw new Error(`${toolName} snapshot ${snapshotId} ended with status=${status}`);
  }
  return getSnapshotData(snapshotId);
}
